src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -772,10 +774,21 @@ impl LocalFingerprint {
     // TODO: This is allowed at this moment. Should figure out if it makes
     // sense if permitting to read env from the config system.
     #[allow(clippy::disallowed_methods)]
-    fn from_env<K: AsRef<str>>(key: K) -> LocalFingerprint {
+    fn from_env<K: AsRef<str>>(
+        key: K,
+        envs: &BTreeMap<String, Option<OsString>>,
+    ) -> LocalFingerprint {
         let key = key.as_ref();
         let var = key.to_owned();
-        let val = env::var(key).ok();
+
+        let val = envs
+            .get(key)
+            .and_then(|opt| {
+                opt.as_ref()
+                    .and_then(|os_str| os_str.clone().into_string().ok())
+            })
+            .or_else(|| env::var(key).ok());
```

Comment:
I find it strange to fallback to `env::var` if the `envs[key]` is non-UTF8.  Maybe we should call `env::var_os` and deal with `into_string` at the end.

Reply:
To be honest, I didn't consider using `env::var_os` instead of `env::var` because I wanted to minimize the impact of the change and avoid affecting other behaviors. If there is an up-to-date value in the current environment variable, then use envs.get(key); otherwise the previous `env::var(key)` is still used.

It looks like the only difference between `env::var` and `env::var_os` is whether the environment variable is a valid Unicode.